distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,969 @@
+    def resolve_address(self, loc):
+        host, port = parse_host_port(loc)
+        return unparse_host_port(ensure_ip(host), port)
+
+    def get_local_address_for(self, loc):
+        host, port = parse_host_port(loc)
+        host = ensure_ip(host)
+        if ":" in host:
+            local_host = get_ipv6(host)
+        else:
+            local_host = get_ip(host)
+        return unparse_host_port(local_host, None)
+
+
+class TLSBackend(TCPBackend):
+    _connector_class = TLSConnector
+    _listener_class = TLSListener
+
+
+# This class is based on parts of `asyncio.selector_events._SelectorSocketTransport`
+# (https://github.com/python/cpython/blob/dc4a212bd305831cb4b187a2e0cc82666fcb15ca/Lib/asyncio/selector_events.py#L757).
+class _ZeroCopyWriter:
+    """The builtin socket transport in asyncio makes a bunch of copies, which
+    can make sending large amounts of data much slower. This hacks around that.
+
+    Note that this workaround isn't used with the windows ProactorEventLoop or
+    uvloop."""
+
+    # We use sendmsg for scatter IO if it's available. Since bookkeeping
+    # scatter IO has a small cost, we want to minimize the amount of processing
+    # we do for each send call. We assume the system send buffer is < 4 MiB
+    # (which would be very large), and set a limit on the number of buffers to
+    # pass to sendmsg.
+    if hasattr(socket.socket, "sendmsg"):
+        try:
+            SENDMSG_MAX_COUNT = os.sysconf("SC_IOV_MAX")
+        except Exception:
+            SENDMSG_MAX_COUNT = 16  # Should be supported on all systems
+    else:
+        SENDMSG_MAX_COUNT = 1  # sendmsg not supported, use send instead
+
+    def __init__(self, protocol, transport):
+        self.protocol = protocol
+        self.transport = transport
+        self._loop = asyncio.get_running_loop()
+
+        # This class mucks with the builtin asyncio transport's internals.
+        # Check that the bits we touch still exist.
+        for attr in [
+            "_sock",
+            "_sock_fd",
+            "_fatal_error",
+            "_eof",
+            "_closing",
+            "_conn_lost",
+            "_call_connection_lost",
+        ]:
+            assert hasattr(transport, attr)
+        # Likewise, this calls a few internal methods of `loop`, ensure they
+        # still exist.
+        for attr in ["_add_writer", "_remove_writer"]:
+            assert hasattr(self._loop, attr)
+
+        # A deque of buffers to send
+        self._buffers: collections.deque[memoryview] = collections.deque()
+        # The total size of all bytes left to send in _buffers
+        self._size = 0
+        # Is the backing protocol paused?
+        self._protocol_paused = False
+        # Initialize the buffer limits
+        self.set_write_buffer_limits()
+
+    def set_write_buffer_limits(self, high: int = None, low: int = None):
+        """Set the write buffer limits"""
+        # Copied almost verbatim from asyncio.transports._FlowControlMixin
```

Comment:
Any reason we couldn't use their implementation as-is?

Reply:
I'd rather not rely on a private class in asyncio, it was easy enough to copy & modify to suit this classes needs.